File: src/exodus/identity.py
"""Node identity: a persistent Ed25519 key pair.

The identity file lives at ``<data_dir>/identity.key`` and is created on first
run.  It is the node's unforgeable name on the network: the public key pins the
node id, and every claim/vote it ever makes is signed with the private key.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

from exodus.crypto import (
    generate_key_pair,
    node_id_from_public_key,
)

PRIVATE_KEY_BYTES = 32
PUBLIC_KEY_BYTES = 32


class IdentityError(RuntimeError):
    pass


@dataclass(frozen=True)
class NodeIdentity:
    node_id: str
    private_key: bytes
    public_key_hex: str

    @property
    def public_key(self) -> bytes:
        return bytes.fromhex(self.public_key_hex)


def load_or_create_identity(path: Path) -> NodeIdentity:
    """Load the identity from *path*, creating it if missing."""

    if path.exists():
        return _load_identity(path)
    return _create_identity(path)


def _create_identity(path: Path) -> NodeIdentity:
    path.parent.mkdir(parents=True, exist_ok=True)
    key_pair = generate_key_pair()
    identity = NodeIdentity(
        node_id=node_id_from_public_key(key_pair.public_key),
        private_key=key_pair.private_key,
        public_key_hex=key_pair.public_key.hex(),
    )
    payload = (
        f"{identity.private_key.hex()}\n"
        f"{identity.public_key_hex}\n"
        f"{identity.node_id}\n"
    )
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    try:
        with os.fdopen(fd, "w") as handle:
            handle.write(payload)
    except OSError:
        if path.exists():
            return _load_identity(path)  # lost a race, use the winner's key
        raise
    return identity


def _load_identity(path: Path) -> NodeIdentity:
    try:
        lines = path.read_text(encoding="utf-8").strip().splitlines()
        private_hex = lines[0].strip()
        public_hex = lines[1].strip()
        private_key = bytes.fromhex(private_hex)
        if len(private_key) != PRIVATE_KEY_BYTES:
            raise IdentityError("bad private key length")
        if len(bytes.fromhex(public_hex)) != PUBLIC_KEY_BYTES:
            raise IdentityError("bad public key length")
    except (OSError, ValueError, IndexError) as exc:
        raise IdentityError(f"cannot read identity file {path}: {exc}") from exc
    return NodeIdentity(
        node_id=node_id_from_public_key(bytes.fromhex(public_hex)),
        private_key=private_key,
        public_key_hex=public_hex,
    )

File: src/identity.rs
//! Node identity: a persistent Ed25519 key pair stored in `<data_dir>/identity.key`.

use std::fs;
use std::os::unix::fs::OpenOptionsExt;
use std::path::Path;

use crate::crypto::{generate_key_pair, hex, hex_decode, node_id_from_public_key};

pub const PRIVATE_KEY_BYTES: usize = 32;
pub const PUBLIC_KEY_BYTES: usize = 32;

#[derive(Debug, Clone)]
pub struct NodeIdentity {
    pub node_id: String,
    pub private_key: Vec<u8>,
    pub public_key_hex: String,
}

impl NodeIdentity {
    pub fn public_key(&self) -> Vec<u8> {
        hex_decode(&self.public_key_hex).expect("stored public key is valid hex")
    }
}

#[derive(Debug)]
pub struct IdentityError(pub String);

impl std::fmt::Display for IdentityError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for IdentityError {}

/// Load the identity from *path*, creating it if missing.
pub fn load_or_create_identity(path: &Path) -> Result<NodeIdentity, IdentityError> {
    if path.exists() {
        load_identity(path)
    } else {
        create_identity(path)
    }
}

fn create_identity(path: &Path) -> Result<NodeIdentity, IdentityError> {
    let parent = path
        .parent()
        .ok_or_else(|| IdentityError(format!("no parent dir for {}", path.display())))?;
    fs::create_dir_all(parent).map_err(|e| IdentityError(e.to_string()))?;

    let (private, public) = generate_key_pair();
    let public_hex = hex(&public);
    let identity = NodeIdentity {
        node_id: node_id_from_public_key(&public),
        private_key: private.to_vec(),
        public_key_hex: public_hex,
    };
    let payload = format!(
        "{}\n{}\n{}\n",
        hex(&identity.private_key),
        identity.public_key_hex,
        identity.node_id
    );

    let mut opts = fs::OpenOptions::new();
    opts.write(true).create_new(true).mode(0o600);
    let mut handle = opts.open(path).map_err(|e| IdentityError(e.to_string()))?;
    use std::io::Write;
    handle
        .write_all(payload.as_bytes())
        .map_err(|e| IdentityError(e.to_string()))?;
    Ok(identity)
}

fn load_identity(path: &Path) -> Result<NodeIdentity, IdentityError> {
    let text = fs::read_to_string(path).map_err(|e| IdentityError(e.to_string()))?;
    let line = |i: usize| text.lines().nth(i).map(|l| l.trim().to_string());
    let private_hex = line(0).ok_or_else(|| IdentityError("malformed identity file".into()))?;
    let public_hex = line(1).ok_or_else(|| IdentityError("malformed identity file".into()))?;

    let private_key = hex_decode(&private_hex).ok_or_else(|| IdentityError("bad private key".into()))?;
    let public_key = hex_decode(&public_hex).ok_or_else(|| IdentityError("bad public key".into()))?;
    if private_key.len() != PRIVATE_KEY_BYTES {
        return Err(IdentityError("bad private key length".into()));
    }
    if public_key.len() != PUBLIC_KEY_BYTES {
        return Err(IdentityError("bad public key length".into()));
    }
    Ok(NodeIdentity {
        node_id: node_id_from_public_key(&public_key),
        private_key,
        public_key_hex: public_hex,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn identity_roundtrip() {
        let dir = std::env::temp_dir().join(format!("exodus-id-test-{}", std::process::id()));
        let path = dir.join("identity.key");
        let id1 = load_or_create_identity(&path).unwrap();
        let id2 = load_or_create_identity(&path).unwrap();
        assert_eq!(id1.node_id, id2.node_id);
        assert_eq!(id1.private_key, id2.private_key);
        assert!(id1.node_id.starts_with("exd"));
        let _ = fs::remove_dir_all(&dir);
    }
}